src/java/org/apache/cassandra/db/tries/Trie.java

```diff
@@ -0,0 +1,596 @@
+    static final ByteComparable.Version BYTE_COMPARABLE_VERSION = ByteComparable.Version.OSS42;
+
+    /**
+     * Adapter interface providing the methods a {@link Walker} to a {@link Consumer}, so that the latter can be used
+     * with {@link #process}.
+     *
+     * This enables calls like
+     *     trie.forEachEntry(x -> System.out.println(x));
+     * to be mapped directly to a single call to {@link #process} without extra allocations.
+     */
+    public interface ValueConsumer<T> extends Consumer<T>, Walker<T, Void>
+    {
+        default void content(T content)
+        {
+            accept(content);
+        }
+
+        default Void complete()
+        {
+            return null;
+        }
+
+        default void resetPathLength(int newDepth)
+        {
+            // not tracking path
+        }
+
+        default void addPathByte(int nextByte)
+        {
+            // not tracking path
+        }
+
+        default void addPathBytes(UnsafeBuffer buffer, int pos, int count)
+        {
+            // not tracking path
+        }
+    }
+
+    /**
+     * Call the given consumer on all content values in the trie in order.
+     */
+    public void forEachValue(ValueConsumer<T> consumer)
+    {
+        process(consumer);
+    }
+
+    /**
+     * Call the given consumer on all (path, content) pairs with non-null content in the trie in order.
+     */
+    public void forEachEntry(BiConsumer<ByteComparable, T> consumer)
+    {
+        process(new TrieEntriesWalker.WithConsumer<T>(consumer));
+        // Note: we can't do the ValueConsumer trick here, because the implementation requires state and cannot be
+        // implemented with default methods alone.
+    }
+
+    /**
+     * Process the trie using the given Walker.
+     */
+    public <R> R process(Walker<T, R> walker)
+    {
+        return process(walker, cursor());
+    }
+
+    static <T, R> R process(Walker<T, R> walker, Cursor<T> cursor)
+    {
+        assert cursor.depth() == 0;
+        T content = cursor.content();   // handle content on the root node
+        if (content == null)
+            content = cursor.advanceToContent(walker);
+
+        while (content != null)
+        {
+            walker.content(content);
+            content = cursor.advanceToContent(walker);
+        }
+        return walker.complete();
+    }
+
+    /**
+     * Constuct a textual representation of the trie.
+     */
+    public String dump()
+    {
+        return dump(Object::toString);
+    }
+
+    /**
+     * Constuct a textual representation of the trie using the given content-to-string mapper.
+     */
+    public String dump(Function<T, String> contentToString)
+    {
+        return process(new TrieDumper<>(contentToString));
+    }
+
+    /**
+     * Returns a singleton trie mapping the given byte path to content.
+     */
+    public static <T> Trie<T> singleton(ByteComparable b, T v)
+    {
+        return new SingletonTrie<>(b, v);
+    }
```

Comment:
nit: I'd consider putting this static factory in `SingletonTrie` itself. Usage could look something like `SingletonTrie.of(b, v)`. Too much knowledge of implementations in the `Trie` interface itself seems worth avoiding.

Reply:
This makes sense in general, but the purpose of this interface is to provide some shared basic functionality, and a singleton trie to me qualifies as such.